### ISSUE TITLE
DB::Exception: Trying to get CSN for too old TID when killing clickhouse server process while big atomic insert

### ISSUE THREAD

**Issue by alsugiliazova** [ClickHouse/ClickHouse#60406]
I create database and table:
```sql
CREATE TABLE IF NOT EXISTS test_table_A (timestamp DateTime,host String,response_time Int32,Path String,Time DateTime,Value Float64,Timestamp Int64,sign Int8)
ENGINE = MergeTree
ORDER BY (host, timestamp)
```

I make a big insert into the table and kill clickhouse server process in parallel:
```sql
BEGIN TRANSACTION;INSERT INTO test_table_A (timestamp, host, response_time, sign) SELECT now() + number/10, toString(number), number, 1 FROM numbers(50000000000); COMMIT;
```
```
kill -SIGKILL {pid}
```
I get:
```
Error on processing query: Code: 32. DB::Exception: Attempt to read after eof: while receiving packet from localhost:9000. (ATTEMPT_TO_READ_AFTER_EOF) (version 23.3.19.32 (official build))
```

I restart server  and get:
```
Code: 210. DB::NetException: Connection refused (localhost:9000). (NETWORK_ERROR)
```

In server error log I get
```
2024.02.26 12:01:18.988795 [ 1880 ] {} <Error> Application: Code: 49. DB::Exception: Trying to get CSN for too old TID (40, 35, 82e71eab-e7ff-4faa-8bb5-16b7bf3e4bfa), current tail_ptr is 41, probably it's a bug: Cannot attach table `test_database5c2304b9_d496_11ee_901a_cf40acb2a173`.`table_A5c2304b9_d496_11ee_901a_cf40acb2a173` from metadata file /var/lib/clickhouse/store/a6a/a6a01912-8d18-4880-9e7a-6803f2f31847/table_A5c2304b9_d496_11ee_901a_cf40acb2a173.sql from query ATTACH TABLE test_database5c2304b9_d496_11ee_901a_cf40acb2a173.table_A5c2304b9_d496_11ee_901a_cf40acb2a173 UUID '30b93009-9076-49d4-b71a-52eff8d377bb' (`timestamp` DateTime, `host` String, `response_time` Int32, `Path` String, `Time` DateTime, `Value` Float64, `Timestamp` Int64, `sign` Int8) ENGINE = MergeTree ORDER BY (host, timestamp) SETTINGS index_granularity = 8192. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0xbb942a4 in /usr/bin/clickhouse
1. ? @ 0x1078c0c8 in /usr/bin/clickhouse
2. DB::TransactionLog::assertTIDIsNotOutdated(DB::TransactionID const&, std::atomic<unsigned long> const*) @ 0x10788c4c in /usr/bin/clickhouse
3. DB::MergeTreeData::loadDataPart(DB::MergeTreePartInfo const&, String const&, std::shared_ptr<DB::IDisk> const&, DB::MergeTreeDataPartState, std::mutex&) @ 0x11002d7c in /usr/bin/clickhouse
4. ? @ 0x11075b80 in /usr/bin/clickhouse
5. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0xbc4dbd0 in /usr/bin/clickhouse
6. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, long, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0xbc5051c in /usr/bin/clickhouse
7. ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0xbc49fa4 in /usr/bin/clickhouse
8. ? @ 0xbc4f448 in /usr/bin/clickhouse
9. start_thread @ 0x7624 in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
10. ? @ 0xd149c in /usr/lib/aarch64-linux-gnu/libc-2.31.so
 (version 23.3.19.32 (official build))
2024.02.26 12:01:18.989006 [ 1880 ] {} <Error> Application: DB::Exception: Trying to get CSN for too old TID (40, 35, 82e71eab-e7ff-4faa-8bb5-16b7bf3e4bfa), current tail_ptr is 41, probably it's a bug: Cannot attach table `test_database5c2304b9_d496_11ee_901a_cf40acb2a173`.`table_A5c2304b9_d496_11ee_901a_cf40acb2a173` from metadata file /var/lib/clickhouse/store/a6a/a6a01912-8d18-4880-9e7a-6803f2f31847/table_A5c2304b9_d496_11ee_901a_cf40acb2a173.sql from query ATTACH TABLE test_database5c2304b9_d496_11ee_901a_cf40acb2a173.table_A5c2304b9_d496_11ee_901a_cf40acb2a173 UUID '30b93009-9076-49d4-b71a-52eff8d377bb' (`timestamp` DateTime, `host` String, `response_time` Int32, `Path` String, `Time` DateTime, `Value` Float64, `Timestamp` Int64, `sign` Int8) ENGINE = MergeTree ORDER BY (host, timestamp) SETTINGS index_granularity = 8192
```
Caught on versions 23.3. and 23.8 on arm machine.

**Comment by alsugiliazova** [comment]
Got the same issue on x86 on latest:
```
(version 25.5.2.47 (official build))
2025.06.21 02:58:14.495115 [ 13207 ] {70ffd6ef-f5f0-4be9-adac-51fa036c04c9} <Error> executeQuery: Code: 722. DB::Exception: Waited job failed: Code: 696. DB::Exception: Load job 'startup table test_databasecde45508_4e3a_11f0_8bdd_960004652619.table_Acde45508_4e3a_11f0_8bdd_960004652619' -> Code: 695. DB::Exception: Load job 'load table test_databasecde45508_4e3a_11f0_8bdd_960004652619.table_Acde45508_4e3a_11f0_8bdd_960004652619' failed: Code: 49. DB::Exception: Trying to get CSN for too old TID (52, 33, d912fe8c-bfca-43f4-b37e-060f048db5e7), current tail_ptr is 53, probably it's a bug: Cannot attach table `test_databasecde45508_4e3a_11f0_8bdd_960004652619`.`table_Acde45508_4e3a_11f0_8bdd_960004652619` from metadata file store/4db/4dbcab91-14be-4b18-a6f3-165a1b9dc718/table_Acde45508_4e3a_11f0_8bdd_960004652619.sql from query ATTACH TABLE test_databasecde45508_4e3a_11f0_8bdd_960004652619.table_Acde45508_4e3a_11f0_8bdd_960004652619 UUID '27f9629c-da14-40dd-8c6a-9536ed28b257' (`timestamp` DateTime, `host` String, `response_time` Int32, `Path` String, `Time` DateTime, `Value` Float64, `Timestamp` Int64, `sign` Int8) ENGINE = MergeTree ORDER BY (host, timestamp) SETTINGS index_granularity = 8192. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000f88f17b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000009da318c
2. DB::Exception::Exception<DB::TransactionID const&, unsigned long&>(int, FormatStringHelperImpl<std::type_identity<DB::TransactionID const&>::type, std::type_identity<unsigned long&>::type>, DB::TransactionID const&, unsigned long&) @ 0x0000000013b5f84b
3. DB::TransactionLog::assertTIDIsNotOutdated(DB::TransactionID const&, std::atomic<unsigned long> const*) @ 0x0000000013b5a735
4. DB::MergeTreeData::loadDataPart(DB::MergeTreePartInfo const&, String const&, std::shared_ptr<DB::IDisk> const&, DB::MergeTreeDataPartState, std::mutex&) @ 0x00000000148b1a6b
5. DB::MergeTreeData::loadDataPartWithRetries(DB::MergeTreePartInfo const&, String const&, std::shared_ptr<DB::IDisk> const&, DB::MergeTreeDataPartState, std::mutex&, unsigned long, unsigned long, unsigned long) @ 0x00000000148b73d4
6. void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::MergeTreeData::loadDataPartsFromDisk(std::vector<std::shared_ptr<DB::MergeTreeData::PartLoadingTree::Node>, std::allocator<std::shared_ptr<DB::MergeTreeData::PartLoadingTree::Node>>>&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000001496ca18
7. void DB::ThreadPoolCallbackRunnerLocal<void, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>, std::function<void ()>>::executeCallback<std::function<void ()>>(std::promise<void>&, std::function<void ()>&&) @ 0x000000001102e6f5
8. DB::ThreadPoolCallbackRunnerLocal<void, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>, std::function<void ()>>::operator()(std::function<void ()>&&, Priority, std::optional<unsigned long>)::'lambda'()::operator()() @ 0x000000001102e4fb
9. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::worker() @ 0x000000000f9cb78b
10. void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000f9d23a3
11. ThreadPoolImpl<std::thread>::ThreadFromThreadPool::worker() @ 0x000000000f9c89b2
12. void* std::__thread_proxy[abi:ne190107]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void (ThreadPoolImpl<std::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::thread>::ThreadFromThreadPool*>>(void*) @ 0x000000000f9cfe7a
13. ? @ 0x000075f85b810ac3
14. ? @ 0x000075f85b8a2850
 (version 25.5.2.47 (official build)). (ASYNC_LOAD_FAILED) (version 25.5.2.47 (official build)). (ASYNC_LOAD_CANCELED) (version 25.5.2.47 (official build)). (ASYNC_LOAD_WAIT_FAILED) (version 25.5.2.47 (official build)) (from 127.0.0.1:36406) (query 1, line 1) (in query: SELECT count(*) + 737 FROM test_databasecde45508_4e3a_11f0_8bdd_960004652619.table_Acde45508_4e3a_11f0_8bdd_960004652619 FORMAT TabSeparated ), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000f88f17b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000009da318c
2. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x0000000009da2eab
3. DB::AsyncLoader::wait(std::shared_ptr<DB::LoadJob> const&, bool) @ 0x000000000fb8a211
4. DB::DatabaseOrdinary::waitTableStarted(String const&) const @ 0x0000000012bee8ee
5. DB::DatabaseWithOwnTablesBase::tryGetTable(String const&, std::shared_ptr<DB::Context const>) const @ 0x0000000012c4a924
6. DB::DatabaseAtomic::tryGetTableUUID(String const&) const @ 0x0000000012bc0930
7. DB::Context::resolveStorageID(DB::StorageID, DB::Context::StorageNamespace) const @ 0x0000000013674277
8. DB::IdentifierResolver::tryResolveTableIdentifier(DB::Identifier const&, std::shared_ptr<DB::Context const> const&) @ 0x00000000131a6038
9. DB::QueryAnalyzer::tryResolveIdentifier(DB::IdentifierLookup const&, DB::IdentifierResolveScope&, DB::IdentifierResolveContext) @ 0x0000000012f6c2a8
10. DB::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::IdentifierResolveScope&) @ 0x0000000012f4f85f
11. DB::QueryAnalyzer::resolve(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::Context const>) @ 0x0000000012f4e52c
12. DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x0000000012f4dd93
13. DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x0000000012fa3e96
14. DB::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) (.llvm.8943257258983934148) @ 0x00000000138ba78d
15. DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000138b871c
16. std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> std::__function::__policy_invoker<std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::registerInterpreterSelectQueryAnalyzer(DB::InterpreterFactory&)::$_0, std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>>(std::__function::__policy_storage const*, DB::InterpreterFactory::Arguments const&) (.llvm.8943257258983934148) @ 0x00000000138bc062
17. DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x000000001386556b
18. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*, std::shared_ptr<DB::IAST>&) @ 0x0000000013c17029
19. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000013c10d26
20. DB::TCPHandler::runImpl() @ 0x0000000014fa4c7e
21. DB::TCPHandler::run() @ 0x0000000014fc39f9
22. Poco::Net::TCPServerConnection::start() @ 0x00000000186e4307
23. Poco::Net::TCPServerDispatcher::run() @ 0x00000000186e4759
24. Poco::PooledThread::run() @ 0x00000000186afa5b
25. Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000186adf3d
26. ? @ 0x000075f85b810ac3
27. ? @ 0x000075f85b8a2850

```

**Comment by tuanpach** [comment]
## Reproduce
1. Add some sleep in `MergeTreeData::loadDataPartWithRetries` to delay the data part loading.
2. Run a very heavy insert in a transaction
```
CREATE TABLE IF NOT EXISTS test_table_A (timestamp DateTime,host String,response_time Int32,Path String,Time DateTime,Value Float64,Timestamp Int64,sign Int8)
ENGINE = MergeTree
ORDER BY (host, timestamp);

BEGIN TRANSACTION;
INSERT INTO test_table_A (timestamp, host, response_time, sign) SELECT now() + number/10, toString(number), number, 1 FROM numbers(50000000000);
COMMIT;
```
3. Create another table
```
CREATE TABLE IF NOT EXISTS test_table_B (timestamp DateTime,host String,response_time Int32,Path String,Time DateTime,Value Float64,Timestamp Int64,sign Int8)
ENGINE = MergeTree
ORDER BY (host, timestamp);
```
4. Insert data into the second table in a transaction
```
BEGIN TRANSACTION;
INSERT INTO test_table_B (timestamp, host, response_time, sign) SELECT now() + number/10, toString(number), number, 1 FROM numbers(5);
COMMIT;
```
5. Kill the clickhouse instance
6. Restart the instance
7. Insert data again into the second table in a transaction
```
BEGIN TRANSACTION;
INSERT INTO test_table_B (timestamp, host, response_time, sign) SELECT now() + number/10, toString(number), number, 1 FROM numbers(5);
COMMIT;
```

## Reason
In the heavy insert transaction, the data parts were not committed before the instance was killed.
After restarting and having a new transaction committed, the `tail_ptr` was updated in `TransactionLog::removeOldEntries` which was larger than the uncommitted data parts `start_csn`. It triggered the assertion while loading the data parts in `MergeTreeData::loadDataPartWithRetries`.

```

0. /repo/ClickHouse/contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x000000001a5db492
1. /repo/ClickHouse/src/Common/Exception.cpp:115: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000ff6f19b
2. /repo/ClickHouse/src/Common/Exception.h:119: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000093331cc
3. /repo/ClickHouse/src/Common/Exception.h:137: DB::Exception::Exception<DB::TransactionID const&, unsigned long&>(int, FormatStringHelperImpl<std::type_identity<DB::TransactionID const&>::type, std::type_identity<unsigned long&>::type>, DB::TransactionID const&, unsigned long&) @ 0x00000000155185eb
4. /repo/ClickHouse/src/Interpreters/TransactionLog.cpp:620: DB::TransactionLog::assertTIDIsNotOutdated(DB::TransactionID const&, std::atomic<unsigned long> const*) @ 0x0000000015513c5a
5. /repo/ClickHouse/src/Interpreters/TransactionLog.cpp:589: DB::TransactionLog::getCSNAndAssert(DB::TransactionID const&, std::atomic<unsigned long>&) @ 0x0000000015517710
6. /repo/ClickHouse/src/Storages/MergeTree/MergeTreeData.cpp:1738: DB::MergeTreeData::loadDataPart(DB::MergeTreePartInfo const&, String const&, std::shared_ptr<DB::IDisk> const&, DB::MergeTreeDataPartState, std::mutex&) @ 0x0000000016ce15b2
7. /repo/ClickHouse/src/Storages/MergeTree/MergeTreeData.cpp:1854: DB::MergeTreeData::loadDataPartWithRetries(DB::MergeTreePartInfo const&, String const&, std::shared_ptr<DB::IDisk> const&, DB::MergeTreeDataPartState, std::mutex&, unsigned long, unsigned long, unsigned long) @ 0x0000000016ce6eda
8. /repo/ClickHouse/src/Storages/MergeTree/MergeTreeData.cpp:1916: void std::__function::__policy_invoker<void ()>::__call_impl[abi:se190107]<std::__function::__default_alloc_func<DB::MergeTreeData::loadDataPartsFromDisk(std::vector<std::shared_ptr<DB::MergeTreeData::PartLoadingTree::Node>, std::allocator<std::shared_ptr<DB::MergeTreeData::PartLoadingTree::Node>>>&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x0000000016d6ea3e
9. /repo/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x000000001138cc90
10. /repo/ClickHouse/src/Common/threadPoolCallbackRunner.h:178: DB::ThreadPoolCallbackRunnerLocal<void, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>, std::function<void ()>>::operator()(std::function<void ()>&&, Priority, std::optional<unsigned long>)::'lambda'()::operator()() @ 0x000000001138cabb
11. /repo/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x00000000100a24f3
12. /repo/ClickHouse/contrib/llvm-project/libcxx/include/__type_traits/invoke.h:117: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'()::operator()() @ 0x00000000100a843d
13. /repo/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x000000001009fb66
14. /repo/ClickHouse/contrib/llvm-project/libcxx/include/__type_traits/invoke.h:117: void* std::__thread_proxy[abi:se190107]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void (ThreadPoolImpl<std::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::thread>::ThreadFromThreadPool*>>(void*) @ 0x00000000100a6000
15. ? @ 0x000000000009caa4
16. ? @ 0x0000000000129c3c
```

**Comment by tavplubix** [comment]
Previously it was working because of https://github.com/ClickHouse/ClickHouse/blob/af5e8b501aa05d45dfd8605fd686c8c653a3dafb/src/Interpreters/TransactionLog.cpp#L269-L276

We were handling parts with unknown commit status during server startup, and either writing their CSNs or rolling them back

But after introducing async parts loading, we can continue loading parts after server startup, and the problem appears. The simplest solution would be to postpone log cleanup until we load (and fix) all parts
